peps/pep-0747.rst

```diff
@@ -0,0 +1,856 @@
+
+Changed signatures
+''''''''''''''''''
+
+The following signatures related to type expressions introduce
+``TypeExpr`` where previously ``object`` existed:
+
+-  ``typing.cast``
+-  ``typing.assert_type``
+
+
+Unchanged signatures
+''''''''''''''''''''
+
+The following signatures related to annotation expressions continue to
+use ``object`` and remain unchanged:
+
+-  ``typing.get_origin``
+-  ``typing.get_args``
+
+The following signatures related to class objects continue to use
+``type`` and remain unchanged:
+
+-  ``builtins.isinstance``
+-  ``builtins.issubclass``
+-  ``builtins.type``
+
+``typing.get_type_hints(..., include_extras=False)`` nearly returns only type
+expressions in Python 3.12, stripping out most type qualifiers
+(``Required, NotRequired, ReadOnly, Annotated``) but currently preserves a
+few type qualifiers which are only allowed in annotation expressions
+(``ClassVar, Final, InitVar, Unpack``). It may be desirable to alter the
+behavior of this function in the future to also strip out those
+qualifiers and actually return type expressions, although this PEP does
+not propose those changes now:
+
+-  ``typing.get_type_hints(..., include_extras=False)``
+
+   -  Almost returns only type expressions, but not quite
+
+-  ``typing.get_type_hints(..., include_extras=True)``
+
+   -  Returns annotation expressions
+
+
+Backwards Compatibility
+=======================
+
+No backward incompatible changes are made by this PEP.
+
+
+How to Teach This
+=================
+
+It is expected that most users interacting with ``TypeExpr`` will do so only
+in a limited way, by passing a literal type expression to a function
+accepting a ``TypeExpr`` input, imported from a runtime type checker
+library.
+
+Runtime type checkers themselves are expected to be the primary in-depth
+users of ``TypeExpr`` values. A function that takes an arbitrary ``TypeExpr`` as
+input must support a large variety of possible type expressions and is
+not easy to write. Some challenges faced by such a function include:
+
+-  An ever-increasing number of typing special forms are introduced with
+   each new Python version which must be recognized, with special
+   handling required for each one.
+-  Stringified type annotations [#strann_less_common]_ (like ``'list[str]'``)
+   must be *parsed* (to something like ``typing.List[str]``) to be introspected.
+
+   -  In practice it is extremely difficult for stringified type
+      annotations to be handled reliably at runtime, so runtime type
+      checkers may opt to not support them at all.
+
+-  Resolving string-based forward references inside type
+   expressions to actual values must typically be done using ``eval()``,
+   which is difficult/impossible to use in a safe way.
+-  Recursive types like ``IntTree = list[typing.Union[int, 'IntTree']]``
+   are not possible to fully resolve.
+-  Supporting user-defined generic types (like Django’s
+   ``QuerySet[User]``) require user-defined functions to
```

Comment:
```suggestion
   ``QuerySet[User]``) requires user-defined functions to
```
Though I'm not sure I fully understand this. Does "user-defined generic" here mean they're not using `typing.Generic`?

Reply:
"User-defined generic" here means that `typing.Generic` is being used.

However many in-the-wild uses of `typing.Generic` don't preserve enough information to check assignability to a generic type at runtime. For example, consider the following type:

```python
class Cell(Generic[T]):
    value: T
    def __init__(self, value: T) -> None:
        self.value = value
```

Given the above definition, how should a general runtime type checker answer the question `isassignable(Cell(1), Cell[int])` at runtime? The type T has been entirely erased, so it's not possible to determine for a given `Cell` instance what value for `T` was given (or inferred) statically.

So a runtime type checker must provide a user-registered hook to recognize `Cell` instances specifically and guess the cell type based on the current value of `.value` for a `Cell`.